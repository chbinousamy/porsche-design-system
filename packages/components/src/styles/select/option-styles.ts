import { isHighContrastMode, Theme } from '../../utils';
import { getHighContrastColors, getThemedColors } from '../colors';
import { borderRadiusSmall, fontLineHeight, spacingStaticSmall } from '../../../../utilities/projects/utilities';
import { getHiddenTextJssStyle, getTransition } from '../common-styles';
import { hoverMediaQuery } from '../hover-media-query';
import { JssStyle, Styles } from 'jss';

export const OPTION_HEIGHT = 40; // optgroups are higher and ignored

export const getSelectOptionStyles = (theme: Theme, additionalOptionJSSStyle?: JssStyle): Styles => {
  const { primaryColor, contrastHighColor, backgroundSurfaceColor, disabledColor } = getThemedColors(theme);
  const { highlightColor } = getHighContrastColors();

  return {
    option: {
      display: 'flex',
      justifyContent: 'space-between',
      gap: '12px',
      padding: `${spacingStaticSmall} 12px`,
      flex: `1 0 calc(${fontLineHeight} + ${spacingStaticSmall} * 2)`,
      color: contrastHighColor,
      cursor: 'pointer',
      textAlign: 'left',
      wordBreak: 'break-word',
      boxSizing: 'border-box',
      borderRadius: borderRadiusSmall,
      transition: ['background-color', 'color'].map(getTransition).join(),
      ...getNoResultsOptionJSSStyles(),
      ...hoverMediaQuery({
        '&:not([aria-disabled]):not([role=status]):hover': {
          color: isHighContrastMode ? highlightColor : primaryColor,
          background: backgroundSurfaceColor,
        },
      }),
      '&--highlighted, &--selected': {
        color: isHighContrastMode ? highlightColor : primaryColor,
        background: backgroundSurfaceColor,
      },
      '&--disabled': {
        cursor: 'not-allowed',
        color: disabledColor,
      },
      '&--hidden': {
        display: 'none',
      },
      ...additionalOptionJSSStyle,
    },
  };
};

export const getNoResultsOptionJSSStyles = (): JssStyle => ({
  '&[role=status]': {
    cursor: 'not-allowed',
  },
  '&__sr': getHiddenTextJssStyle(),
});
