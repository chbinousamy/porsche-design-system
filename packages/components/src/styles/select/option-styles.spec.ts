import { getSelectOptionStyles } from './option-styles';

describe('getSelectOptionStyles()', () => {
  it.each<Parameters<typeof getSelectOptionStyles>>([
    ['light', undefined],
    ['dark', undefined],
    ['light', { color: 'black' }],
    ['dark', { color: 'black' }],
  ])('should return correct css for theme: %s and additionalOptionJSSStyle: %o', (...args) => {
    expect(getSelectOptionStyles(...args)).toMatchSnapshot();
  });
});
