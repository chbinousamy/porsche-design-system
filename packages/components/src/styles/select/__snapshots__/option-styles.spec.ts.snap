// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSelectOptionStyles() should return correct css for theme: dark and additionalOptionJSSStyle: { color: 'black' } 1`] = `
{
  "option": {
    "&--disabled": {
      "color": "#7E7F82",
      "cursor": "not-allowed",
    },
    "&--hidden": {
      "display": "none",
    },
    "&--highlighted": {
      "background": "#404044",
    },
    "&--highlighted, &--selected": {
      "color": "#FBFCFF",
    },
    "&--selected": {
      "background": "#212225",
    },
    "&[role=status]": {
      "cursor": "not-allowed",
    },
    "&__sr": {
      "height": "1px",
      "position": "absolute",
      "textIndent": "-999999px",
      "width": 0,
    },
    "@media(hover:hover)": {
      "&:not([aria-disabled]):not([role=status]):hover": {
        "background": "#404044",
        "color": "#FBFCFF",
      },
    },
    "borderRadius": "4px",
    "boxSizing": "border-box",
    "color": "black",
    "cursor": "pointer",
    "display": "flex",
    "flex": "1 0 calc(calc(6px + 2.125ex) + 8px * 2)",
    "gap": "12px",
    "justifyContent": "space-between",
    "padding": "8px 12px",
    "textAlign": "left",
    "transition": "background-color var(--p-transition-duration, .24s) ease,color var(--p-transition-duration, .24s) ease",
    "wordBreak": "break-word",
  },
}
`;

exports[`getSelectOptionStyles() should return correct css for theme: dark and additionalOptionJSSStyle: undefined 1`] = `
{
  "option": {
    "&--disabled": {
      "color": "#7E7F82",
      "cursor": "not-allowed",
    },
    "&--hidden": {
      "display": "none",
    },
    "&--highlighted": {
      "background": "#404044",
    },
    "&--highlighted, &--selected": {
      "color": "#FBFCFF",
    },
    "&--selected": {
      "background": "#212225",
    },
    "&[role=status]": {
      "cursor": "not-allowed",
    },
    "&__sr": {
      "height": "1px",
      "position": "absolute",
      "textIndent": "-999999px",
      "width": 0,
    },
    "@media(hover:hover)": {
      "&:not([aria-disabled]):not([role=status]):hover": {
        "background": "#404044",
        "color": "#FBFCFF",
      },
    },
    "borderRadius": "4px",
    "boxSizing": "border-box",
    "color": "#AFB0B3",
    "cursor": "pointer",
    "display": "flex",
    "flex": "1 0 calc(calc(6px + 2.125ex) + 8px * 2)",
    "gap": "12px",
    "justifyContent": "space-between",
    "padding": "8px 12px",
    "textAlign": "left",
    "transition": "background-color var(--p-transition-duration, .24s) ease,color var(--p-transition-duration, .24s) ease",
    "wordBreak": "break-word",
  },
}
`;

exports[`getSelectOptionStyles() should return correct css for theme: light and additionalOptionJSSStyle: { color: 'black' } 1`] = `
{
  "option": {
    "&--disabled": {
      "color": "#949598",
      "cursor": "not-allowed",
    },
    "&--hidden": {
      "display": "none",
    },
    "&--highlighted": {
      "background": "#D8D8DB",
    },
    "&--highlighted, &--selected": {
      "color": "#010205",
    },
    "&--selected": {
      "background": "#EEEFF2",
    },
    "&[role=status]": {
      "cursor": "not-allowed",
    },
    "&__sr": {
      "height": "1px",
      "position": "absolute",
      "textIndent": "-999999px",
      "width": 0,
    },
    "@media(hover:hover)": {
      "&:not([aria-disabled]):not([role=status]):hover": {
        "background": "#D8D8DB",
        "color": "#010205",
      },
    },
    "borderRadius": "4px",
    "boxSizing": "border-box",
    "color": "black",
    "cursor": "pointer",
    "display": "flex",
    "flex": "1 0 calc(calc(6px + 2.125ex) + 8px * 2)",
    "gap": "12px",
    "justifyContent": "space-between",
    "padding": "8px 12px",
    "textAlign": "left",
    "transition": "background-color var(--p-transition-duration, .24s) ease,color var(--p-transition-duration, .24s) ease",
    "wordBreak": "break-word",
  },
}
`;

exports[`getSelectOptionStyles() should return correct css for theme: light and additionalOptionJSSStyle: undefined 1`] = `
{
  "option": {
    "&--disabled": {
      "color": "#949598",
      "cursor": "not-allowed",
    },
    "&--hidden": {
      "display": "none",
    },
    "&--highlighted": {
      "background": "#D8D8DB",
    },
    "&--highlighted, &--selected": {
      "color": "#010205",
    },
    "&--selected": {
      "background": "#EEEFF2",
    },
    "&[role=status]": {
      "cursor": "not-allowed",
    },
    "&__sr": {
      "height": "1px",
      "position": "absolute",
      "textIndent": "-999999px",
      "width": 0,
    },
    "@media(hover:hover)": {
      "&:not([aria-disabled]):not([role=status]):hover": {
        "background": "#D8D8DB",
        "color": "#010205",
      },
    },
    "borderRadius": "4px",
    "boxSizing": "border-box",
    "color": "#535457",
    "cursor": "pointer",
    "display": "flex",
    "flex": "1 0 calc(calc(6px + 2.125ex) + 8px * 2)",
    "gap": "12px",
    "justifyContent": "space-between",
    "padding": "8px 12px",
    "textAlign": "left",
    "transition": "background-color var(--p-transition-duration, .24s) ease,color var(--p-transition-duration, .24s) ease",
    "wordBreak": "break-word",
  },
}
`;
